### PR TITLE
BUG: SparseDataFrame construction with lists not coercing to dtype  (GH 15682)

### LINKED ISSUE #15682: BUG: SparseDataFrame construction with lists not coercing to dtype
coercion happens
```
In [12]: df = pd.SparseDataFrame({'a': [1, 0, 0],
    ...:                          'b': [0, 1, 0],
    ...:                          'c': [0, 0, 1]}, dtype='uint8', default_fill_value=0)

In [13]: df.dtypes
Out[13]: 
a    int64
b    int64
c    int64
dtype: object
```

respected here
```
In [23]: df = pd.SparseDataFrame({'a': pd.SparseSeries([1, 0, 0], dtype='uint8'),
    ...:                          'b': pd.SparseSeries([0, 1, 0], dtype='uint8'),
    ...:                          'c': pd.SparseSeries([0, 0, 1], dtype='uint8')}, default_fill_value=0)

In [24]: df.dtypes
Out[24]: 
a    uint8
b    uint8
c    uint8
dtype: object

In [25]: df.default_fill_value
Out[25]: 0
```

And here
```
In [15]: df = pd.DataFrame({'a': [1, 0, 0],
    ...:                          'b': [0, 1, 0],
    ...:                          'c': [0, 0, 1]}, dtype='uint8').to_sparse(fill_value=0)

In [17]: df.dtypes
Out[17]: 
a    uint8
b    uint8
c    uint8
dtype: object

In [18]: df.default_fill_value
Out[18]: 0
```

### PR DESCRIPTION
**Pull request by ucals**
- [x] closes #15682
 - [x] tests added / passed
 - [ ] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/sparse/frame.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def _init_dict(self, data, index, columns, dtype=None):
                     v = [v.get(i, nan) for i in index]
 
                 v = sp_maker(v)
-            sdict[k] = v
+            sdict[k] = v.astype(np.dtype(dtype)) if dtype is not None else v
```
Comment (jreback) [inline pandas/sparse/frame.py]: you would do this in ``sp_maker``, also don't use if/else on a single line (hard to read)

File: pandas/tests/sparse/test_frame.py [hunk 2]
```diff
@@ -758,7 +757,8 @@ def test_sparse_frame_fillna_limit(self):
     def test_rename(self):
         # just check this works
         renamed = self.frame.rename(index=str)  # noqa
-        renamed = self.frame.rename(columns=lambda x: '%s%d' % (x, len(x)))  # noqa
```
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: can you compare these results (as no tests before).

File: pandas/tests/sparse/test_frame.py [hunk 3]
```diff
@@ -1226,7 +1225,6 @@ def test_from_to_scipy_object(spmatrix, fill_value):
 
 
 class TestSparseDataFrameArithmetic(tm.TestCase):
-
```
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: try to turn off this removing blank lines like this

File: pandas/tests/sparse/test_frame.py [hunk 4]
```diff
@@ -1299,3 +1296,12 @@ def test_numpy_func_call(self):
                  'std', 'min', 'max']
         for func in funcs:
             getattr(np, func)(self.frame)
+
+    def test_type_coercion_at_construction(self):
+        # GH 15682
+        df = pd.SparseDataFrame(
+            {'a': [1, 0, 0], 'b': [0, 1, 0], 'c': [0, 0, 1]}, dtype='uint8',
+            default_fill_value=0)
+        result = df.dtypes[0]
```
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: create the result and use ``assert_sp_frame_equal``

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/sparse/test_frame.py (hunk 4 above)
Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: you can use the comparison from the issue (IOW construct both ways and compare)

Comment (jreback) [inline pandas/tests/sparse/test_frame.py]: move this test where other ``test_constructor`` tests are

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/sparse/test_frame.py (hunk 4 above)
Comment (ucals) [inline pandas/tests/sparse/test_frame.py]: done

File: pandas/sparse/frame.py (hunk 1 above)
Comment (ucals) [inline pandas/sparse/frame.py]: done

File: pandas/tests/sparse/test_frame.py (hunk 2 above)
Comment (ucals) [inline pandas/tests/sparse/test_frame.py]: didn't understand

File: pandas/tests/sparse/test_frame.py (hunk 3 above)
Comment (ucals) [inline pandas/tests/sparse/test_frame.py]: just put it back, sorry

File: pandas/tests/sparse/test_frame.py (hunk 2 above)
Comment (ucals) [inline pandas/tests/sparse/test_frame.py]: ok, just figured it out, added the right test on my last commit! :)

**Pull request closed without merging**